File: Algorithms/Python/Unique-Paths-II.py
class Solution(object):

    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        if not obstacleGrid or obstacleGrid[0][0] == 1 or obstacleGrid[-1][-1] == 1:
            return 0
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        ans = [[0 for x in range(n)] for x in range(m)]
        flag = False
        for i in range(m):
            if obstacleGrid[i][0] == 1:
                flag = True
            ans[i][0] = 0 if flag else 1
        flag = False
        for j in range(n):
            if obstacleGrid[0][j] == 1:
                flag = True
            ans[0][j] = 0 if flag else 1
        for i in range(1, m):
            for j in range(1, n):
                ans[i][j] = 0 if obstacleGrid[i][
                    j] == 1 else ans[i - 1][j] + ans[i][j - 1]
        return ans[m - 1][n - 1]

File: Algorithms/Java/Unique Paths II.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
		int n = obstacleGrid[0].length;
		int[][] num = new int[m][n];
		for (int i = 0; i < m; i++) {
			if (obstacleGrid[i][0] != 1)
				num[i][0] = 1;
			else
				break;
		}
		for (int j = 0; j < n; j++) {
			if (obstacleGrid[0][j] != 1)
				num[0][j] = 1;
			else
				break;
		}
		for (int i = 1; i < m; i++) {
			for (int j = 1; j < n; j++) {
				if (obstacleGrid[i][j] == 1)
					num[i][j] = 0;
				else
					num[i][j] = num[i - 1][j] + num[i][j - 1];
			}
		}
		return num[m - 1][n - 1];
    }
}